src/plugins/usage_collection/server/collector/collector.ts

```diff
@@ -17,7 +17,30 @@
  * under the License.
  */
 
-export class Collector {
+import { Logger } from 'kibana/server';
+import { CallCluster } from 'src/legacy/core_plugins/elasticsearch';
+
+export type CollectorFormatForBulkUpload<T, U> = (result: T) => { type: string; payload: U };
+
+export interface CollectorOptions<T = unknown, U = T> {
+  type: string;
+  init?: Function;
+  fetch: (callCluster: CallCluster) => Promise<T> | T;
```

Comment:
I believe `callCluster` should be optional

```
  fetch: (callCluster?: CallCluster) => Promise<T> | T;
```

There is no need to pass generic `T` at the interface level. We do not need the fetch return type when registering the collector. It could be scoped to fetch or simply the use of `..) => unknown` is enough.

Reply:
> I believe `callCluster` should be optional

If we do it optional, every fetch will need to check if `callCluster` is not `undefined` before using it, but from the `CollectorSet` we are always providing it (and we should be enforced to provide it).

Declaring your `fetch` method without any argument if not using the `callCluster` parameter is still fine for TS (and even encouraged by `eslint`). A quick TS-safe example I just tested below:
```typescript
  usageCollection.makeUsageCollector({
    type: KIBANA_REPORTING_TYPE,
    isReady,
    fetch: () => ({ count: 1 }),
  });
```

> There is no need to pass generic T at the interface level. We do not need the fetch return type when registering the collector. It could be scoped to fetch or simply the use of ..) => unknown is enough.

The reason for being at the interface level is so `formatForBulkUpload` infers the type from the output of `fetch`. And it also will fail to compile if `formatForBulkUpload` is assuming an input different to the output of `fetch`.

An example is [`x-pack/legacy/plugins/reporting/server/usage/reporting_usage_collector.ts`](https://github.com/afharo/kibana/blob/telemetry/os-distro/x-pack/legacy/plugins/reporting/server/usage/reporting_usage_collector.ts#L25-L49). You can see there is no need to specify the types when calling `makeUsageCollector` as it infers it _automagically_ from the methods definitions.